src/Datadog.Trace/ServiceFabric/ServiceRemotingHelpers.cs

```diff
@@ -0,0 +1,176 @@
+            string spanKind,
+            IServiceRemotingRequestEventArgs? eventArgs,
+            IServiceRemotingRequestMessageHeader? messageHeader)
+        {
+            string? methodName = null;
+            string? resourceName = null;
+            string? serviceUrl = null;
+
+            string serviceFabricServiceName = PlatformHelpers.ServiceFabric.ServiceName;
+
+            if (eventArgs != null)
+            {
+                methodName = eventArgs.MethodName ??
+                             messageHeader?.MethodName ??
+                             messageHeader?.MethodId.ToString(CultureInfo.InvariantCulture) ??
+                             "unknown_method";
+
+                serviceUrl = eventArgs.ServiceUri?.AbsoluteUri;
+                resourceName = serviceUrl == null ? methodName : $"{serviceUrl}/{methodName}";
+            }
+
+            var tags = new ServiceRemotingTags(spanKind)
+            {
+                ApplicationId = PlatformHelpers.ServiceFabric.ApplicationId,
+                ApplicationName = PlatformHelpers.ServiceFabric.ApplicationName,
+                PartitionId = PlatformHelpers.ServiceFabric.PartitionId,
+                NodeId = PlatformHelpers.ServiceFabric.NodeId,
+                NodeName = PlatformHelpers.ServiceFabric.NodeName,
+                ServiceName = serviceFabricServiceName,
+                RemotingUri = serviceUrl,
+                RemotingMethodName = methodName
+            };
+
+            if (messageHeader != null)
+            {
+                tags.RemotingMethodId = messageHeader.MethodId.ToString(CultureInfo.InvariantCulture);
+                tags.RemotingInterfaceId = messageHeader.InterfaceId.ToString(CultureInfo.InvariantCulture);
+                tags.RemotingInvocationId = messageHeader.InvocationId;
+            }
+
+            Span span = tracer.StartSpan(GetSpanName(spanKind), tags, context);
+            span.ResourceName = resourceName;
+            tags.SetAnalyticsSampleRate(ServiceRemotingConstants.IntegrationId, Tracer.Instance.Settings, enabledWithGlobalSetting: false);
+
+            return span;
+        }
+
+        public static void FinishSpan(EventArgs? e, string spanKind)
+        {
+            try
+            {
+                var scope = Tracer.Instance.ActiveScope;
+
+                if (scope == null)
+                {
+                    Log.Warning("Expected an active scope, but there is none.");
+                    return;
+                }
+
+                string expectedSpanName = GetSpanName(spanKind);
+
+                if (expectedSpanName != scope.Span.OperationName)
+                {
+                    Log.Warning("Expected span name {expectedSpanName}, but found {actualSpanName} instead.", expectedSpanName, scope.Span.OperationName);
+                    return;
+                }
+
+                try
+                {
+                    var eventArgs = e?.DuckAs<IServiceRemotingFailedResponseEventArgs>();
```

Comment:
On successful requests, the `DuckAs` call does not throw an exception and the `eventArgs` variable is assigned `null`, right?

Reply:
Yes. `As<T>()` used to throw, but @tonyredondo refactored and renamed the duck typing extension methods to align better with the `as` and `is` C# operators. See https://github.com/DataDog/dd-trace-dotnet/pull/1220